File: algorithms/Graphs/Medium/KrushkalMST/VSDKrushkalMST.py



# Function to convert adjacency matrix to adjacency list
def graph(g):
    x=[]
    for i in range(len(g)):
        for j in range(len(g[0])):
            if g[i][j]!=0:
                x.append([i,j,g[i][j]])
    return x


# Function to find the parent of given vertex
def find(parent,i):
    if parent[i]==i:
        return i
    return find(parent,parent[i])


# Function to apply union of given two subtrees
def union(parent,rank,x,y):
    xroot=find(parent,x)
    yroot=find(parent,y)
    if rank[xroot]<rank[yroot]:
        parent[xroot]=yroot
    elif rank[xroot]>rank[yroot]:
        parent[yroot]=xroot
    else:
        parent[yroot]=xroot
        rank[xroot]+=1



# Function to find MST of given graph 
def VSDkruskal(graph,n):
        result = []
        i=0
        e = 0
        graph = sorted(graph, key=lambda item: item[2])
        parent = []
        rank = []
        for node in range(n):
            parent.append(node)    
            rank.append(0)
        while e < n - 1:
            u, v, w = graph[i]
            i = i + 1
            x = find(parent, u)
            y = find(parent, v)
            if x != y:
                e = e + 1
                result.append([u, v, w])
                union(parent, rank, x, y)
        cost=0
        print("-"*30)
        print("The Kruskal's MST Graph is :")
        for u, v, weight in result:
            cost+=weight
            print(f"{u} - {v} : {weight}")
        print("The total cost of Krushkal's MST is",cost)


# The Driver Function
if __name__=="__main__":
    n=5
    g =[ [ 0, 2, 0, 6, 0 ],
        [ 2, 0, 3, 8, 5 ],
        [ 0, 3, 0, 0, 7 ],
        [ 6, 8, 0, 0, 9 ],
        [ 0, 5, 7, 9, 0 ] 
    ]   
    x=graph(g)
    VSDkruskal(x,n)

File: algorithms/Graphs/Medium/KrushkalMST/VSDKrushkalMST.java
/*Copyrights to vsdevelopers.io*/
/*For more programs visit vsdevelopers.io */
//Java program for finding Minimum Spanning tree from given graph using Kruskal's algorithm
import java.util.LinkedList;

public class VSDKrushkalMST {
	//Class to represent the structure of the edge along with weights
	public static class VSDEdge {
		 int source;
		 int destination;
		 int weight;

		 public VSDEdge(int source, int destination, int weight) {
		 this.source = source;
		 this.destination = destination;
		 this.weight = weight;
		 }
		 }
	//Class to represent the structure of the graph	
public static class VSDGraph{
	int v;//No.of vertices
	LinkedList<VSDEdge> adList[];//Adjacency list to mark the edges
	//Initializing vertices using constructor
	VSDGraph(int size){
		v=size;
		adList=new LinkedList[v];
		for(int i=0;i<v;i++)
			adList[i]= new LinkedList<VSDEdge>();
	}
	//Function to add edge to the graph
	public void VSDaddEgde(int source, int destination, int weight) {
		 VSDEdge edge = new VSDEdge(source, destination, weight);
		 adList[source].addLast(edge); // adding new edge to the list
		 }
	//Function to implement prim's algorithm
	public  void VSDKruskalsMst(VSDGraph g) {
		int edgecount=0;
		int weight=0;//To calculate the weight of MST.
		 boolean[] visited = new boolean[g.v];//Array to mark the vertices alredy present in MST
		    for(int i=0;i<g.v;i++) {
		    	visited[i]=false;
		    }
		    System.out.println("The Kruskal's MST is:");
		    while (edgecount<g.v-1) {
		    int min=999999;
		    int x = 0; // source vertex
		    int y = 0; // destination vertex
		    for(int i=0;i<g.v;i++) {
			   LinkedList<VSDEdge> list = adList[i];
			   for(int j=0;j<list.size();j++) {
				   if(!visited[list.get(j).destination]&&min > list.get(j).weight)
				   {
					   min=list.get(j).weight;
					   x=i;
					   y=list.get(j).destination;
				   }
			   }
		   }
		   System.out.println(x+" "+y);
		   if(edgecount==0) visited[x] = true;
		   edgecount++;
		   weight+=min;
		   visited[y] = true;
		    }
		    System.out.println("The total cost of Kruskal's MST is "+ weight);
	}
	
	
}

public static void main(String args[]) {
	VSDGraph g=new VSDGraph(6);
	g.VSDaddEgde(0,1,20);
	g.VSDaddEgde(1,0,20);
	g.VSDaddEgde(0,2,30);
	g.VSDaddEgde(2,0,30);
	g.VSDaddEgde(0,3,40);
	g.VSDaddEgde(3,0,40);
	g.VSDaddEgde(0,4,20);
	g.VSDaddEgde(4,0,20);
	g.VSDaddEgde(1,2,10);
	g.VSDaddEgde(2,1,10);
	g.VSDaddEgde(1,3,5);
	g.VSDaddEgde(3,1,5);
	g.VSDaddEgde(1,5,10);
	g.VSDaddEgde(5,1,10);
	g.VSDaddEgde(2,3,20);
	g.VSDaddEgde(3,2,20);
	g.VSDaddEgde(2,4,50);
	g.VSDaddEgde(4,2,50);
	g.VSDaddEgde(2,5,20);
	g.VSDaddEgde(5,2,20);
	g.VSDaddEgde(3,5,40);
	g.VSDaddEgde(5,3,40);
	g.VSDaddEgde(4,5,30);
	g.VSDaddEgde(5,4,30);
	g.VSDKruskalsMst(g);
}
}
